pdns/gss_context.cc

```diff
@@ -0,0 +1,588 @@
+      buffer.value = const_cast<void*>(static_cast<const void*>(name.c_str()));
+      OM_uint32 min;
+      auto maj = gss_import_name(&min, &buffer, (gss_OID)GSS_KRB5_NT_PRINCIPAL_NAME, &d_name);
+      if (maj != GSS_S_COMPLETE) {
+        d_name = GSS_C_NO_NAME;
+        d_valid = false;
+        return;
+      }
+    }
+
+    renew();
+  };
+
+  ~GssCredential()
+  {
+    OM_uint32 tmp_maj __attribute__((unused)), tmp_min __attribute__((unused));
+    if (d_cred != GSS_C_NO_CREDENTIAL) {
+      tmp_maj = gss_release_cred(&tmp_min, &d_cred);
+    }
+    if (d_name != GSS_C_NO_NAME) {
+      tmp_maj = gss_release_name(&tmp_min, &d_name);
+    }
+  };
+
+  bool expired() const
+  {
+    if (d_expires == -1) {
+      return false;
+    }
+    return time(nullptr) > d_expires;
+  }
+
+  bool renew()
+  {
+    OM_uint32 time_rec, tmp_maj __attribute__((unused)), tmp_min __attribute__((unused));
+    tmp_maj = gss_acquire_cred(&tmp_min, d_name, GSS_C_INDEFINITE, GSS_C_NO_OID_SET, d_usage, &d_cred, nullptr, &time_rec);
+
+    if (tmp_maj != GSS_S_COMPLETE) {
+      d_valid = false;
+      tmp_maj = gss_release_name(&tmp_min, &d_name);
+      d_name = GSS_C_NO_NAME;
+      return false;
+    }
+
+    d_valid = true;
+
+    // We do not want forever, but a good time
+    if (time_rec == GSS_C_INDEFINITE) {
+      time_rec = 24 * 60 * 60;
+    }
+    d_expires = time(nullptr) + time_rec;
+
+    return true;
+  }
+
+  bool valid()
+  {
+    return d_valid && !expired();
+  }
+
+  std::string d_nameS;
+  gss_cred_usage_t d_usage;
+  gss_name_t d_name{GSS_C_NO_NAME};
+  gss_cred_id_t d_cred{GSS_C_NO_CREDENTIAL};
+  time_t d_expires{time(nullptr) + 60}; // partly initialized wil be cleaned up
+  bool d_valid{false};
+}; // GssCredential
+
+static LockGuarded<std::unordered_map<std::string, std::shared_ptr<GssCredential>>> s_gss_accept_creds;
+static LockGuarded<std::unordered_map<std::string, std::shared_ptr<GssCredential>>> s_gss_init_creds;
+
+class GssSecContext : boost::noncopyable
+{
+public:
+  GssSecContext(std::shared_ptr<GssCredential> cred) :
+    d_cred(cred)
+  {
+    if (!cred->valid()) {
+      throw PDNSException("Invalid credential " + cred->d_nameS);
+    }
+    d_cred = cred;
+    d_state = GssStateInitial;
+    d_ctx = GSS_C_NO_CONTEXT;
+    d_expires = 0;
+    d_peer_name = GSS_C_NO_NAME;
+    d_type = GSS_CONTEXT_NONE;
+  }
+
+  ~GssSecContext()
+  {
+    OM_uint32 tmp_maj __attribute__((unused)), tmp_min __attribute__((unused));
+    if (d_ctx != GSS_C_NO_CONTEXT) {
+      tmp_maj = gss_delete_sec_context(&tmp_min, &d_ctx, GSS_C_NO_BUFFER);
+    }
+    if (d_peer_name != GSS_C_NO_NAME) {
+      tmp_maj = gss_release_name(&tmp_min, &(d_peer_name));
+    }
+  }
+
+  std::shared_ptr<GssCredential> d_cred;
+  GssContextType d_type{GSS_CONTEXT_NONE};
+  gss_ctx_id_t d_ctx{GSS_C_NO_CONTEXT};
+  gss_name_t d_peer_name{GSS_C_NO_NAME};
+  time_t d_expires{time(nullptr) + 60}; // partly initialized wil be cleaned up
```

Comment:
is this different from `d_expires = 0` in the constructor on purpose?

Reply:
This was a bug indeed. I cganged the constructor to not overwrite `d_expires` (and use the init values for the other fields as well).